src/emqx_delayed_publish.erl

```diff
@@ -100,6 +83,26 @@ handle_call({store, DelayedMsg = #delayed_message{key = Key}}, _From, State) ->
     ok = mnesia:dirty_write(?TAB, DelayedMsg),
     {reply, ok, ensure_publish_timer(Key, State)};
 
+handle_call({cancel_publish, ClientId}, _From, State = #{timer := TRef}) ->
+    case mnesia:dirty_select(?TAB, [{#delayed_message{client_id = ClientId, _ = '_'}, [], ['$_']}]) of
+        [] -> 
+            {reply, ok, State};
+        WillMsgs ->
+            Result = lists:keysearch(mnesia:dirty_first(?TAB), 2, WillMsgs),
```

Comment:
Change the mnesia type from `set` to `ordered_set`, or this `first .. next` iteration on a set ets table will not work as expected, as set tables are not sorted in the order of insertion.